src/server/tiering/external_alloc.cc

```diff
@@ -384,8 +396,13 @@ void ExternalAllocator::Free(size_t offset, size_t sz) {
   ++page->available;
 
   DCHECK_EQ(page->available, page->free_blocks.count());
+  // If page becomes fully free, return it to segment list, otherwise if it just became non-empty,
+  // then return it to free pages list
   if (page->available == blocks_num) {
     FreePage(page, seg, block_size);
+  } else if (page->available == 1) {
+    page->next_free = free_pages_[page->bin_idx];
```

Comment:
note, this may introduce a jerking behaviour where a page with a single block will be restored,removed and soon (no action items).